examples/ssh/sshcert/sshcert.go

```diff
@@ -0,0 +1,160 @@
+package sshcert
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+
+	// "github.com/openpubkey/openpubkey/parties"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/openpubkey/openpubkey/client"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	"golang.org/x/crypto/ssh"
+)
+
+func NewSshSignerFromPem(pemBytes []byte) (ssh.MultiAlgorithmSigner, error) {
+	rawKey, err := ssh.ParseRawPrivateKey(pemBytes)
+	if err != nil {
+		return nil, err
+	}
+	sshSigner, err := ssh.NewSignerFromKey(rawKey)
+	if err != nil {
+		return nil, err
+	}
+	return ssh.NewSignerWithAlgorithms(sshSigner.(ssh.AlgorithmSigner), []string{ssh.KeyAlgoRSASHA256})
+}
+
+type SshCertSmuggler struct {
+	SshCert *ssh.Certificate
+}
+
+func New(pkt *pktoken.PKToken, principals []string) (*SshCertSmuggler, error) {
+
+	// TODO: assumes email exists in ID Token,
+	// this will break for OPs like Azure that do not have email as a claim
+	var claims struct {
+		Email string `json:"email"`
+	}
+	if err := json.Unmarshal(pkt.Payload, &claims); err != nil {
+		return nil, err
+	}
+
+	pubkeySsh, err := sshPubkeyFromPKT(pkt)
+	if err != nil {
+		return nil, err
+	}
+	pktJson, err := pkt.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	pktB64 := string(util.Base64EncodeForJWT(pktJson))
+	sshSmuggler := SshCertSmuggler{
+		SshCert: &ssh.Certificate{
+			Key:             pubkeySsh,
+			CertType:        ssh.UserCert,
+			KeyId:           claims.Email,
+			ValidPrincipals: principals,
+			ValidBefore:     ssh.CertTimeInfinity,
+			Permissions: ssh.Permissions{
+				Extensions: map[string]string{
+					"permit-X11-forwarding":   "",
+					"permit-agent-forwarding": "",
+					"permit-port-forwarding":  "",
+					"permit-pty":              "",
+					"permit-user-rc":          "",
+					"openpubkey-pkt":          pktB64,
+				},
+			},
+		},
+	}
+	return &sshSmuggler, nil
+}
+
+func NewFromAuthorizedKey(certType string, certB64 string) (*SshCertSmuggler, error) {
+	if certPubkey, _, _, _, err := ssh.ParseAuthorizedKey([]byte(certType + " " + certB64)); err != nil {
+		return nil, err
+	} else {
+		opkcert := &SshCertSmuggler{
+			SshCert: certPubkey.(*ssh.Certificate),
+		}
+		return opkcert, nil
+	}
+}
+
+func (s *SshCertSmuggler) SignCert(signerMas ssh.MultiAlgorithmSigner) (*ssh.Certificate, error) {
```

Comment:
this variable name should just be `signer`

Reply:
I want to call this out as a special signer. Look at the code that calls it:

```
sshSigner, err := ssh.NewSignerFromSigner(signer)
	if err != nil {
		return nil, nil, err
	}

	signerMas, err := ssh.NewSignerWithAlgorithms(sshSigner.(ssh.AlgorithmSigner), []string{ssh.KeyAlgoECDSA256})
	if err != nil {
		return nil, nil, err
	}

	sshCert, err := cert.SignCert(signerMas)
```

Reply:
a functional argument name shouldn't be defined by the code that calls it because it's existence should be predicated on its use by various code sources. `signerMas` is unnecessary and obviously I hate the name. You're even sending an `ssh.AlgorithmSigner` and then treating it as a `ssh.MultiAlgorithmSigner` which I'm sure would make sense if I went to go read crypto.ssh but I'd rather not have our code rely on a user's knowledge of that

Reply:
but anyways, feel free to resolve, it's not worth arguing about this